cleanup coverage flags, upload coverage report

sarnold opened the pull request:
* speedup: switch coverage workflow to ctest
* cleanup: make sure covrage builds are identical
* add cov report artifact upload to smoke workflow (no gh-pages branch yet)
* add/update coverage cfgs and tox commands

<!---
Thank you for your contribution. Please make sure your pull request fulfills
all of the requirements below. If you cannot currently tick all the boxes,
but would still like to create a PR, then add the label "WIP" and assign
the PR to yourself.

If your PR fixes a bug, include the regression test(s) in the same commit
as the bug fix. Else, keep commits small and orthogonal, possibly placing
tests in commits of their own.
--->

- [x] Each commit message has a non-empty body, explaining why the change was made.
- [x] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [x] The feature or user visible behavior I have added or modified has been documented in the README file.
- [x] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [x] My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- [x] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

<!---
See `.gitchangelog.rc` in the top-level repo directory for guidance on
commit message summary formatting and tags.  Please use the appropriate
action and/or audience markers in the first line of your commit messages.

See also, e.g., https://chris.beams.io/posts/git-commit/ for general
guidelines on commit messages.

If you have created commits mixing multiple features and/or unrelated
white-space changes, use git rebase or a sequence involving git reset
and git add -p to fix this.
--->

Commit:
fix: dev: cleanup coverage flags, upload coverage report

* speedup: switch coverage workflow to ctest
* cleanup: make sure covrage builds are identical
* add cov report artifact upload to smoke workflow (no gh-pages branch yet)
* add/update coverage cfgs and tox commands

Signed-off-by: Stephen L Arnold <nerdboy@gentoo.org>

Commit:
fix: dev: cleanup ci cmds (per OS env), add python dep for conda devenv

* bionic lcov is too old for required include usage
* devenv needs jinja python dep per CI version

Signed-off-by: Stephen L Arnold <nerdboy@gentoo.org>

SJLC commented:
Are you going to create a gh-pages branch? Or, is there some other way to see the uploaded reports?

sarnold commented:
Right now you can only see the html report locally:

```
tox -e auto
epiphany coverage/index.html
```
Sadly lcov `demangle-cpp` does *not* run c++filt on the symbol names, so I was going to look at the gcovr html output at some point soon-ish...  Right now we're just using gcovr for console display/xml report.  Things to remember:

* accurate coverage needs 2 things
  * no optimizing away code paths/branches (generation)
  * proper "filtering" for only the needed source files (reporting)
* each tool should produce the same values for lines/branches/funcs from both builds
  * note there are trivial/small differences in the html report between gcovr and lcov; eg, gcovr does not report on a line with only a closing brace, but in at least one case lcov *does* report such a line as "uncovered"
  * also note we use only gcovr for human/CI coverage display/analysis (except for the lcov html report)
* lcov html can show coverage for lines/branches *or* lines/functions
* gcovr html/text can only show coverage for lines/branches
* the more "actionable" data is the function/branch coverage (hits/misses, not covered, etc)

Commit:
fix: dev: create both html coverage reports, add inc/ dir to metrics artifact

* create both html reports, one for functions and one for branches
* note each report is created via separate tox cmds
* sync up metrics source code with coverage
* update conda devenv file, use Ninja generator

Signed-off-by: Stephen L Arnold <nerdboy@gentoo.org>

chg: dev: py38/39 is not resolving deps like 37, remove jinja py ver

* this should really not be necessary, somehow devenv is inconsistent
* it should work fine across all python versions 36 => 39
* even on macos

Signed-off-by: Stephen L Arnold <nerdboy@gentoo.org>

chg: dev: set base conda env to py 3.x and restore jinja ver

* set default to 37 instead of 39, reset to macos-11

Signed-off-by: Stephen L Arnold <nerdboy@gentoo.org>

Commit:
chg: dev: json_test coughs runtime exception on macos, disable for now

* tested on macos-10/11 in github (macos-10 did not run)

Signed-off-by: Stephen L Arnold <nerdboy@gentoo.org>

Commit:
chg: doc: add section on tox commands, remove redis-ipc-py section

Signed-off-by: Stephen L Arnold <nerdboy@gentoo.org>
